### PR TITLE
[GNA] Fix for concat layer with >2 inputs

### PR DESCRIPTION
**Pull request by bsochack**
Signed-off-by: Bartosz Sochacki <bartosz.sochacki@intel.com>

### CODE REVIEW

File: inference-engine/src/gna_plugin/layers/gna_layer_info.hpp [hunk 1]
```diff
@@ -260,7 +260,10 @@ class LayerInfo {
         auto cropLayer = dynamic_cast<InferenceEngine::CropLayer *> (layer);
         if (cropLayer != nullptr && !cropLayer->offset.empty()) {
             try {
-                size_t cropOffset = cropLayer->offset.back() * cropLayer->precision.size();
+                // currently crop layer only supports 2 bytes in int16 and int8 mode.
+                // In fp32 mode this is not necessary but is useful for testing
+                auto bytesPerCropElement = 2;
+                size_t cropOffset = cropLayer->offset.back() * bytesPerCropElement;
                 return (ALIGN64(cropOffset) != cropOffset);
             } catch (InferenceEngine::details::InferenceEngineException) {}
```
Comment (esmirno) [inline inference-engine/src/gna_plugin/layers/gna_layer_info.hpp:268]: wonder what kind of exception we intercepting here - suggest remove try-catch at all

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 2]
```diff
@@ -782,8 +778,9 @@ void InsertConcatAligningFilterPass::run() {
                 size_t num_rows_out = num_rows_padded + num_rows_in;
 
                 // encodes offset to beginning of split layer input
+                size_t off = (aligned64_offset / bytesPerConcatElement) * (quantized ? bytesPerConcatElement : 4);
```
Comment (esmirno) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:781]: rename off to something useful like bytesOffset

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 3]
```diff
@@ -837,84 +834,92 @@ void ReorderConcatInputsPass::run() {
     }
     int numOfLinkLayers = 0;
 
-    for (auto & l : *pLayers) {
-        // 1st stage locate concat align filter
+    for (auto& l : *pLayers) {
+        // 1st stage locate concat
         LayerInfo info(l);
-        if (!info.isConcatAlignFilter()) {
+        if (!info.isConcat()) {
```
Comment (esmirno) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:840]: the whole purpose of align filters reordering looks misunderstood - for concat without overlap reodering not needeed at all any execution order should work.

**Review by esmirno — CHANGES_REQUESTED** [review]
dont know what we are fixing here - reodering mechanics will be extra constrained for some reason

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 3 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:840]: please go through the code more carefully. The first step is to find concat layer.
In the second step checking all inputs to concat layer to find concat align ones.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 2 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:781]: ok

File: inference-engine/src/gna_plugin/layers/gna_layer_info.hpp (hunk 1 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/layers/gna_layer_info.hpp:268]: ok

File: inference-engine/src/gna_plugin/gna_graph_compiler.cpp [hunk 4]
```diff
@@ -112,6 +112,12 @@ void GNAGraphCompiler::fillConcatConnections(InferenceEngine::CNNLayerPtr layer)
                 InferenceEngine::details::product(begin(dataInput->getDims()),
                                                   end(dataInput->getDims())) * dataInput->getPrecision().size();
 
+        // concat align layer can have additional padding, so the size of layer needs to be calculated
+        // based on original number of rows
+        if (ptrConcatLayerInput->CheckParamPresence("original_num_rows")) {
```
Comment (bsochack) [inline inference-engine/src/gna_plugin/gna_graph_compiler.cpp:117]: @esmirno please take a look at this change and let me know if it is acceptable.

Comment (esmirno) [reply inference-engine/src/gna_plugin/gna_graph_compiler.cpp:117]: this looks OK

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 5]
```diff
@@ -837,84 +834,92 @@ void ReorderConcatInputsPass::run() {
...
             continue;
         }
 
-        // 2rd locating concat
-        if (l->outData.size() != 1) {
-            THROW_GNA_EXCEPTION << "no concat layer after concat aligning layer" << l->name;
+        // 2nd stage locate first input in concat
```
Comment (esmirno) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:844]: lets restrict to fp32 only, so >= 2 can work in that way , integer i would not declare as this mode supported

Comment (bsochack) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:850]: I have pushed a fix for it. please review

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp [hunk 6]
```diff
@@ -382,119 +382,148 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
...
-        }
 
-        int concatIdxToUpdate = -1;
+        // support only cases when all inputs have the same quant value
+        auto inputLayerCheck = [](const auto& inputLayer) {
+            auto info = LayerInfo(inputLayer);
+            return info.isInput();
+        };
+
+        GNAPluginNS::QuantizedLayerParams* sourceQuantParams = NULL;
```
Comment (dorloff) [inline inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:421]: Why not nullptr?

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp [hunk 7]
```diff
@@ -382,119 +382,148 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
...
 
-        if (infoIn0.isInput()) {
-            sourceQuantParams = quantParams0;
-        } else if (infoIn1.isInput()) {
-            concatIdxToUpdate = 0;
-            sourceQuantParams = quantParams1;
+                auto quantParamsSecond = InferenceEngine::getInjectedData<QuantizedLayerParams>(*inIt);
+                if (!fp32eq(quantParamsSecond->_dst_quant.scale, quantParamsFirst->_dst_quant.scale)) {
+                    THROW_GNA_EXCEPTION << "Two Input layers " << (*firstInputIt)->name
+                        << "and" << (*inIt)->name << " has different scales in concat!!! \n";
```
Comment (dorloff) [inline inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:433]: "have"

File: inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/concat_multi_input.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
+// Copyright (C) 2019 Intel Corporation
```
Comment (dorloff) [inline inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/concat_multi_input.cpp:1]: 2020

File: inference-engine/tests/functional/plugin/shared/include/subgraph_tests/concat_multi_input.hpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2019 Intel Corporation
```
Comment (dorloff) [inline inference-engine/tests/functional/plugin/shared/include/subgraph_tests/concat_multi_input.hpp:1]: 2020

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/concat_multi_input.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
+// Copyright (C) 2019 Intel Corporation
```
Comment (dorloff) [inline inference-engine/tests/functional/plugin/shared/src/subgraph_tests/concat_multi_input.cpp:1]: 2020

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 11]
```diff
@@ -843,84 +840,92 @@ void ReorderConcatInputsPass::run() {
...
+
+            CNNNetDFS(currConcatLayer, [&currConcatLayer, &prevInputToConcat, &bFinish](CNNLayerPtr layer) {
+                gnalog() << "[UFS] from : " << currConcatLayer->name << " reached: " << layer->name << "\n";
+                // found that direct input to concat is a indirect parent of align filter - so no link required
+                if (layer.get() == prevInputToConcat.get() || LayerInfo(prevInputToConcat).isInput()) {
+                    gnalog() << "[UFS] copy layer insertion needed\n";
+                    bFinish = true;
+                }
+                }, true, [&bFinish](InferenceEngine::CNNLayer* from) {
+                    // aborting UFS once link not need
```
Comment (dorloff) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:902]: "not needed"

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp [hunk 12]
```diff
@@ -382,119 +382,148 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
...
-                          return;
-                      }
-                      restartedLayer = layer;
-                      gnalog() << "... OK,  need requantize\n";
-                  }, true, [&restartedLayer, &concatLayer, &concatIdxToUpdate](InferenceEngine::CNNLayer *from) {
+        for (auto& layerIdToUpdate : concatIdxToUpdate) {
+            auto destinationQuantParams = InferenceEngine::getInjectedData<QuantizedLayerParams>(*concatLayer);
+            destinationQuantParams->_dst_quant.scale = sourceQuantParams->_dst_quant.scale;
+
+            InferenceEngine::CNNLayerPtr  restartedLayer;
```
Comment (dorloff) [inline inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:492]: extra blank

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp (hunk 6 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp]: copy paste issue. fixed

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp (hunk 7 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp]: copy paste issue. fixed

File: inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/concat_multi_input.cpp (hunk 8 above)
Comment (bsochack) [reply inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/concat_multi_input.cpp]: copy paste issue. fixed

File: inference-engine/tests/functional/plugin/shared/src/subgraph_tests/concat_multi_input.cpp (hunk 10 above)
Comment (bsochack) [reply inference-engine/tests/functional/plugin/shared/src/subgraph_tests/concat_multi_input.cpp]: copy paste issue. fixed

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 11 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp]: copy paste issue. fixed

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp (hunk 12 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:492]: copy paste issue. fixed

File: inference-engine/tests/functional/plugin/shared/include/subgraph_tests/concat_multi_input.hpp (hunk 9 above)
Comment (bsochack) [reply inference-engine/tests/functional/plugin/shared/include/subgraph_tests/concat_multi_input.hpp]: copy paste issue. fixed

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 5 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:850]: please review the latest changes, I proposed a fix for scale factor propagation

File: inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/concat_multi_input.cpp [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+const std::vector<InferenceEngine::Precision> netPrecisions = {
+    InferenceEngine::Precision::FP32,
+    InferenceEngine::Precision::FP16,
+};
+
+std::map<std::string, std::string> additional_config = {
+    {"GNA_COMPACT_MODE", "NO"},
+    {"GNA_SCALE_FACTOR_0", "2048"},
```
Comment (esmirno) [inline inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/concat_multi_input.cpp:30]: for integer mode computation on gna you need to you config keys:

    {"GNA_DEVICE_MODE", "GNA_SW_EXACT"},
    {InferenceEngine::GNAConfigParams::KEY_GNA_PRECISION, "I16"},

Comment (bsochack) [reply inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/concat_multi_input.cpp:30]: done, added

Comment (esmirno) [reply inference-engine/tests/functional/plugin/gna/shared_tests_instances/subgraph_tests/concat_multi_input.cpp:30]: great - how about GNA_DEVICE_MODE? is that set somewhere else - not here?

**Review by esmirno — APPROVED** [review]

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp [hunk 14]
```diff
@@ -382,119 +382,154 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
...
+        auto sourceLayerCheck = [&fp32eq](InferenceEngine::CNNLayerPtr& inputLayer) {
+            auto quantParams = InferenceEngine::getInjectedData<QuantizedLayerParams>(inputLayer);
+            LayerInfo info(inputLayer);
+            return !info.isActivation() && !fp32eq(quantParams->_dst_quant.scale, 1.0f);
+        };
 
-            sourceQuantParams = quantParams0;
-        }
+        auto sourceLayerIt = std::find_if(inputLayers.begin(), inputLayers.end(), sourceLayerCheck);
+        if (sourceLayerIt == inputLayers.end()) {
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:445]: Chosing always the same input layer for scale factor quantization may lead to requantization loop in some cases (it is also the case with current code before your changes) please add some mechanism to change selected input for requantization if this layer is being requantized repeatedly (you can check PR #2080 where map with counter is used for that case, likely not great solution but working)

Comment (bsochack) [reply inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:457]: I have pushed a fix for it.

**Pull request merged**